virtool/http/auth.py

```diff
@@ -102,11 +116,76 @@ async def authenticate_with_api_key(req: Request, handler: Callable, user_id: st
     return await handler(req)
 
 
+async def authenticate_with_b2c(req: Request, handler: Callable, app: Application) -> Response:
+    """
+    Authenticate requests when req.app["config"].use_b2c is True.
+
+    If no id_token cookie is attached to request, redirect to /api/acquire_tokens
+
+    If id_token cookie is found, attempt to validate to gather user information from claims. If token is expired,
+    redirect to /api/refresh_tokens. If token is invalid for some other reason, redirect to /api/delete_tokens
+
+    find or create user based on token claims, then populate req["cient"] with user information and return the
+    response from the handler.
+
+    :param req: the request to handle
+    :param handler: the handler to call with the request if it is authenticated successfully
+    :param app: the application object
+
+    :return: the response
+    """
+    if req.cookies.get("id_token") is None:
+        auth_code_flow = await req.app["run_in_thread"](
+            req.app["msal"].initiate_auth_code_flow,
+            ["email"],
+            f"http://localhost:9950/api/acquire_tokens",
+            random_alphanumeric(8)
+        )
+        req.app["auth_code_flow"] = auth_code_flow
+        return HTTPFound(auth_code_flow["auth_uri"])
+
+    else:
+        try:
+            token_claims = await validate_token(req.cookies.get("id_token"), req.app)
+        except ExpiredSignatureError:
+            return HTTPFound("/api/refresh_tokens")
+        except JWTClaimsError:
+            return HTTPFound("/api/delete_tokens")
+
+        user_attributes = ADUserAttributes(
+            display_name=token_claims["name"],
+            given_name=token_claims["given_name"],
+            family_name=token_claims["family_name"],
+            oid=token_claims["oid"]
+        )
+
+        user_document = await find_or_create_b2c_user(app["db"], user_attributes)
+
+        req["client"] = UserClient(
+            db=app["db"],
+            administrator=user_document["administrator"],
+            force_reset=False,
+            groups=user_document["groups"],
+            permissions=user_document["permissions"],
+            user_id=user_document["_id"],
+            authenticated=True
+        )
+
+        return await handler(req)
+
+
 @web.middleware
-async def middleware(req, handler):
+async def middleware(req, handler) -> Response:
+    """
+    Handle requests based on client type and authentication status.
+
+    :param req: the request to handle
+    :param handler: the handler to call with the request if it is authenticated successfully
+    :return: the response
+    """
     db = req.app["db"]
 
-    if req.path == "/api/account/login" or req.path == "/api/account/logout":
+    if req.path in UNAUTHENTICATED_ROUTES:
```

Comment:
This should be doable with `public` as well. I'm not sure why explicit checking of `req.path` was done here.